pkg/azuredisk/controllerserver.go

```diff
@@ -839,6 +844,36 @@ func (d *Driver) getSnapshotByID(ctx context.Context, resourceGroup, snapshotID,
 	return generateCSISnapshot(sourceVolumeID, &snapshot)
 }
 
+// GetSourceDiskSize returns: disk size, error
+func (d *Driver) GetSourceDiskSize(resourceGroup, diskName string) (*int32, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	return d.GetSourceDiskSizeHelper(ctx, resourceGroup, diskName)
+}
+
+// GetSourceDiskSizeHelper is a helper function for GetSourceDiskSize to support recursive search for  and it returns: disk size, error
+func (d *Driver) GetSourceDiskSizeHelper(ctx context.Context, resourceGroup, diskName string) (*int32, error) {
+	result, rerr := d.cloud.DisksClient.Get(ctx, resourceGroup, diskName)
+
+	if rerr != nil {
+		return nil, rerr.Error()
+	}
+
+	if result.DiskProperties.CreationData != nil && (*result.DiskProperties.CreationData).CreateOption == "Copy" {
+		klog.V(2).Infof("Clone source disk has a parent source")
+		sourceResourceID := *result.DiskProperties.CreationData.SourceResourceID
+		parentResourceGroup, _ := GetResourceGroupFromURI(sourceResourceID)
+		parentDiskName := path.Base(sourceResourceID)
+		return d.GetSourceDiskSizeHelper(ctx, parentResourceGroup, parentDiskName)
```

Comment:
if current disk has parent disk, why return parent disk size? should it always return current disk size?

Reply:
It is a safety measure. As I was marking through a few test cases, I observed that if you clone a new target volume from an previously cloned volume that has a parent disk of smaller size, the new target volume's filesystem inherits the parent disk's fs capacity instead of cloned volume's fs capacity.

Without the recursive search, you can replicate the error by:
```
Clone a 16Gi target volume B from a 8Gi volume A.
B's target capacity > A's capacity
     B's filesystem resized in nodeserver
B's filesystem shows 16Gi capacity

Clone a 16Gi target volume C from B
C's target capacity == B's capacity
     C's filesystem not resized
C's filesystem ends up 8Gi for some reason.
```